server/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzer.java

```diff
@@ -22,73 +22,126 @@
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.analysis.TokenStream;
 import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.analysis.CustomAnalyzerProvider.AnalyzerComponents;
 
 import java.io.Reader;
+import java.util.Map;
 
 public final class CustomAnalyzer extends Analyzer {
+    private CloseableThreadLocal<AnalyzerComponents> storedComponents = new CloseableThreadLocal<>();
+    private volatile AnalyzerComponents current;
 
-    private final String tokenizerName;
-    private final TokenizerFactory tokenizerFactory;
-
-    private final CharFilterFactory[] charFilters;
+    private final AnalysisMode analysisMode;
 
-    private final TokenFilterFactory[] tokenFilters;
+    private static final ReuseStrategy UPDATE_STRATEGY = new ReuseStrategy() {
+        @Override
+        public TokenStreamComponents getReusableComponents(Analyzer analyzer, String fieldName) {
+            CustomAnalyzer custom = (CustomAnalyzer) analyzer;
+            AnalyzerComponents components = custom.getStoredComponents();
+            if (components == null || custom.shouldReload(components)) {
+                custom.setStoredComponents(custom.current);
+                return null;
+            }
+            TokenStreamComponents tokenStream = (TokenStreamComponents) getStoredValue(analyzer);
+            assert tokenStream != null;
+            return tokenStream;
+        }
 
-    private final int positionIncrementGap;
-    private final int offsetGap;
-    private final AnalysisMode analysisMode;
+        @Override
+        public void setReusableComponents(Analyzer analyzer, String fieldName, TokenStreamComponents tokenStream) {
+            setStoredValue(analyzer, tokenStream);
+        }
+    };
 
     public CustomAnalyzer(String tokenizerName, TokenizerFactory tokenizerFactory, CharFilterFactory[] charFilters,
-            TokenFilterFactory[] tokenFilters) {
-        this(tokenizerName, tokenizerFactory, charFilters, tokenFilters, 0, -1);
+                          TokenFilterFactory[] tokenFilters) {
+        this(new AnalyzerComponents(tokenizerName, tokenizerFactory, charFilters, tokenFilters, -1, -1));
```

Comment:
@jimczi going over your commit I wonder about the reason to change the `positionIncrementGap` from 0 to -1? I don't know if it makes a difference or not, just curious.

Reply:
Good catch, this is not intended `0` should be kept.